tsdb/querier.go

```diff
@@ -438,28 +427,31 @@ func labelValuesWithMatchers(ctx context.Context, r IndexReader, name string, ma
 		return nil, errors.Wrap(err, "fetching postings for matchers")
 	}
 
-	allValues, err := r.LabelValues(ctx, name)
-	if err != nil {
-		return nil, errors.Wrapf(err, "fetching values of label %s", name)
-	}
-
 	// If we have a matcher for the label name, we can filter out values that don't match
 	// before we fetch postings. This is especially useful for labels with many values.
 	// e.g. __name__ with a selector like {__name__="xyz"}
+	var filter func(_ string) bool
 	for _, m := range matchers {
-		if m.Name != name {
-			continue
-		}
-
-		// re-use the allValues slice to avoid allocations
-		// this is safe because the iteration is always ahead of the append
-		filteredValues := allValues[:0]
-		for _, v := range allValues {
-			if m.Matches(v) {
-				filteredValues = append(filteredValues, v)
+		if m.Name == name {
+			if filter == nil {
+				filter = m.Matches
+			} else {
+				prevFilter := filter
+				filter = func(s string) bool {
+					// Chain together previous filter with this one.
+					return prevFilter(s) && m.Matches(s)
+				}
 			}
 		}
-		allValues = filteredValues
+	}
+
+	if filter == nil {
+		filter = func(_ string) bool { return true }
+	}
+
+	allValues, err := r.LabelValuesFiltered(ctx, name, filter)
```

Comment:
I think that it would make more sense to have a separate path for `nil` filter where function isn't called and the slice is pre-allocated instead of letting it grow.